proposals/hasCrossSiteAncestor.md

```diff
@@ -0,0 +1,154 @@
+
+## Motivation
+https://github.com/privacycg/CHIPS/issues/40 is adding a cross-site ancestor bit value to partitioned cookies.
+
+### Objective
+To interact with partitioned cookies containing a cross-site ancestor chain bit correctly, extensions will need to have the ability to to specify a value (`hasCrossSiteAncestor`) that corresponds to the value of the cross-site ancestor chain bit in partitioned cookies.
+
+#### Use Cases
+
+#### Cookie Manager:
+Let's say a cookie manager extension (with host permissions) is used by users to get/set/remove their cookies. As browsers include the cross-site ancestor chain bit in their implementation of partitioned cookies, the extension will need the ability to use the `hasCrossSiteAncestor` parameter to give full insight into the existing cookies and allow the user to set new cookies that include the cross-site ancestor bit correctly.
+
+#### Password Manager:
+Let’s say a password manager extension (with host permissions) is used by users to access their login information by setting a cookie that stores their usernames and passwords in an encrypted partitioned cookie. To protect their users against clickjacking, the extension adds a setting that prevents their cookies from being accessed, by default, in embeds that have cross site ancestors without triggering a user prompt. If permission is given through the prompt, the extension sets a cookie with a `hasCrossSiteAncestor` value of true. Upon subsequent visits, the extension checks the cookie store for the presence of a cookie with a `hasCrossSiteAncestor` value of true to determine whether the prompt needs to be rendered.
+
+To allow for this protection and UX flow to work, the extension would need to have the ability to set/get cookies with specific `hasCrossSiteAncestor` values. 
+
+### Known Consumers
+All extensions that access and/or modify cookies with awareness of partitioned cookies, through the use of the `partitionKey` property in the `cookies` extension API.
+
+## Specification
+
+### Schema
+
+Adds a new boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the `cookies.Cookie` type.
+
+Adds a new optional boolean property `hasCrossSiteAncestor` to the `partitionKey` property of the details object. Which is the parameter used in the methods `cookies.get()`, `cookies.set()`, `cookies.remove()` and `cookies.getAll()`.
+
+```
+// Cookie details object with partitionKey containing hasCrossSiteAncestor value of true.
+"details": {
+  "partitionKey": {
+    "topLevelSite": "https://example.com",
+    "hasCrossSiteAncestor": true
+  }
+}
+```
+#### Cookies.GetPartitionKey()
+Adds a new API, `cookies.getPartitionKey()` which retrives a valid `partitionKey` for the indicated frame.
+
+This is an asyncronus function that returns a promise.
+
+##### Syntax
+```
+let key = cookies.getPartitionKey(
+    details // object
+)
+```
+##### Parameters
```

Comment:
(continues thread from https://github.com/w3c/webextensions/pull/581#discussion_r1703979196)

I see two distinct use cases for the getPartitionKey method:
- to get the PartitionKey that applies to cookies with the `Partitioned` attribute in the frame. This is currently covered by the method.
- to know which partition to use when Partitioned is not specified. In Chrome the default is currently to not partition, in Firefox cookies are partitioned if third-party and SSA is not used.

To cover the second use case, we could add another option that specifies the intended scenario. An alternative is to add a method that returns whether the frame can access unpartitioned cookies.

Reply:
I think [hasStorageAccess](https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess) should cover the second use case already.